server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,232 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
+		return err
+	}
+	ul.req.Requests = append(ul.req.Requests, &repb.BatchUpdateBlobsRequest_Request{
+		Digest: d,
+		Data:   b,
+	})
+	ul.currentBatchSize += d.GetSizeBytes()
+	return nil
+}
+
+func (ul *BatchCASUploader) UploadProto(in proto.Message) (*repb.Digest, error) {
+	data, err := proto.Marshal(in)
+	if err != nil {
+		return nil, err
+	}
+	d, err := digest.Compute(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	if err := ul.Upload(d, NewBytesReadSeekCloser(data)); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+func (ul *BatchCASUploader) UploadFile(path string) (*repb.Digest, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	d, err := digest.Compute(f)
+	if err != nil {
+		return nil, err
+	}
+	// Note: uploader.Upload will close the file.
+	if err := ul.Upload(d, f); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+type bytesReadSeekCloser struct {
+	io.ReadSeeker
+}
+
+func NewBytesReadSeekCloser(b []byte) io.ReadSeekCloser {
+	return &bytesReadSeekCloser{bytes.NewReader(b)}
+}
+func (*bytesReadSeekCloser) Close() error { return nil }
+
+func (ul *BatchCASUploader) flushCurrentBatch() {
+	req := ul.req
+	ul.req = &repb.BatchUpdateBlobsRequest{InstanceName: ul.instanceName}
+	ul.currentBatchSize = 0
+	ul.eg.Go(func() error {
+		rsp, err := ul.casClient.BatchUpdateBlobs(ul.ctx, req)
+		if err != nil {
+			return err
+		}
+		for _, fileResponse := range rsp.GetResponses() {
+			if fileResponse.GetStatus().GetCode() != int32(codes.OK) {
+				return gstatus.Error(codes.Code(fileResponse.GetStatus().GetCode()), fmt.Sprintf("Error uploading file: %v", fileResponse.GetDigest()))
+			}
+		}
+		return nil
+	})
+}
+
+func (ul *BatchCASUploader) Wait() error {
+	if len(ul.req.GetRequests()) > 0 {
+		ul.flushCurrentBatch()
+	}
+	return ul.eg.Wait()
+}
+
+// UploadDirectoryToCAS uploads all the files in a given directory to the CAS
+// as well as the directory structure, and returns the digest of the root
+// Directory proto that can be used to fetch the uploaded contents.
+func UploadDirectoryToCAS(ctx context.Context, env environment.Env, instanceName, rootDirPath string) (*repb.Digest, error) {
+	ul, err := NewBatchCASUploader(ctx, env, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	dirs := []*repb.Directory{}
+	// Recursively find and upload all descendent dirs.
+	var rootDirectoryDigest *repb.Digest
+	if rootDirectoryDigest, _, err = uploadDir(ul, rootDirPath, &dirs); err != nil {
+		return nil, err
+	}
+	// Upload the tree, which consists of the root dir as well as all descendant
+	// dirs.
+	rootTree := &repb.Tree{Root: dirs[0], Children: dirs[1:]}
+	if _, err := ul.UploadProto(rootTree); err != nil {
+		return nil, err
+	}
+	if err := ul.Wait(); err != nil {
+		return nil, err
+	}
+	return rootDirectoryDigest, nil
+}
+
+func uploadDir(ul *BatchCASUploader, dirPath string, acc *[]*repb.Directory) (*repb.Digest, *repb.Directory, error) {
```

Comment:
slices are already references so there's no need to do *[]*repb.Directory

Reply:
I was treating `*[]*repb.Directory` as a C-style return value holder (returning the output value of `acc` "into" that pointer), to avoid adding another `nil` when returning an error. Re-wrote to be more idiomatic (just using a regular return value).